android/TritonClient/tritonlib/src/main/java/com/mobiledgex/tritonlib/ImageSenderTriton.java

```diff
@@ -0,0 +1,280 @@
+                    Log.e(TAG, message);
+                    mImageServerInterface.reportConnectionError(error.toString(), ImageSenderTriton.this);
+                }
+            }) {
+                byte[] concatenateByteArrays(byte[] a, byte[] b) {
+                    byte[] result = new byte[a.length + b.length];
+                    System.arraycopy(a, 0, result, 0, a.length);
+                    System.arraycopy(b, 0, result, a.length, b.length);
+                    return result;
+                }
+
+                @Override
+                public byte[] getBody() {
+                    return concatenateByteArrays(requestBody.getBytes(), bytes);
+                }
+
+                @Override
+                public Map getHeaders() {
+                    Map headers = new HashMap();
+                    headers.put("Inference-Header-Content-Length", headerLength+"");
+                    return headers;
+                }
+            };
+
+            // Add the request to the RequestQueue.
+            mRequestQueue.add(mStringRequestMain);
+        } else {
+            Log.e(TAG, "Unknown communication mode: "+ mConnectionMode);
+        }
+    }
+
+    /**
+     * Both the REST server and the GRPC server will return results in the same JSON
+     * format. This method parses the results and updates the UI with the returned values.
+     *
+     * @param response
+     * @param latency
+     */
+    private void handleResponse(String response, long latency) {
+        Log.i(TAG, "handleResponse mModelName="+mModelName+" response="+response);
+        if (mInactive) {
+            Log.i(TAG, "Inactive, aborting update.");
+            return;
+        }
+        try {
+            if (mModelName.equals("ensemble_dali_yolov4")) {
+
+                // Get just the outputs/OBJECTS_JSON data
+                JSONObject jsonObject = new JSONObject(response);
+                JSONArray outputs = jsonObject.getJSONArray("outputs");
+                for (int i = 0; i < outputs.length(); i++) {
+                    JSONObject output = outputs.getJSONObject(i);
+                    String name = output.getString("name");
+                    if (!name.equals("OBJECTS_JSON")) {
+                        continue;
+                    }
+                    JSONArray data = output.getJSONArray("data");
+                    String object = data.getString(0);
+                    Log.i(TAG, "handleResponse object="+object);
+                    response = object.toString();
+                    jsonObject = new JSONObject(response);
+                    JSONArray objects = jsonObject.getJSONArray("objects");
+                    mImageServerInterface.updateOverlay(mCloudLetType, objects, null);
+                }
+
+            } else if (mModelName.equals("ensemble_dali_inception")) {
+                Log.i(TAG, "ensemble_dali_inception response="+response);
+                String inference = response.substring(239);
```

Comment:
I am sure you have a reason for using a substring(constant)
 Is deserializing ensemble_dali_inception JSON response a future step or not needed!

Reply:
I'm glad you asked about this, @ah1053. This was something I hard-coded, but stupidly forgot about. The 239 value can change, so I am probably printing some incorrect "confidence" values sometimes. Here's an example response: `{"model_name":"ensemble_dali_inception","model_version":"1","parameters":{"sequence_id":0,"sequence_start":false,"sequence_end":false},"outputs":[{"name":"OUTPUT","datatype":"BYTES","shape":[1,1],"parameters":{"binary_data_size":22}}]}0.233588:707:PATIO`
I'm supposed to be able get the length from the 'Inference-Header-Content-Length' header, but I was having trouble accessing that data, so hard-coded a value I learned from testing, meaning to revisit it for a proper fix.

Time to revisit it!

Reply:
Upon further review, the Inference-Header-Content-Length is always 235. There are 4 binary characters that we ignore before getting to the actual inference value, so 239 is the right value. I'll clarify in the code where this value comes from.